Chunked array docs (#7951)
pydata/xarray: doc/internals/chunked-arrays.rst

.. currentmodule:: xarray

.. _internals.chunkedarrays:

Alternative chunked array types
===============================

.. warning::

    This is a *highly* experimental feature. Please report any bugs or other difficulties on `xarray's issue tracker <https://github.com/pydata/xarray/issues>`_.
    In particular see discussion on `xarray issue #6807 <https://github.com/pydata/xarray/issues/6807>`_

Xarray can wrap chunked dask arrays (see :ref:`dask`), but can also wrap any other chunked array type that exposes the correct interface.
This allows us to support using other frameworks for distributed and out-of-core processing, with user code still written as xarray commands.

> In particular xarray can now also supports wrapping :py:class:`cubed.Array` objects

Reviewer: ```suggestion
In particular xarray also supports wrapping :py:class:`cubed.Array` objects
```

.. currentmodule:: xarray

.. _internals.chunkedarrays:

Alternative chunked array types
===============================

.. warning::

    This is a *highly* experimental feature. Please report any bugs or other difficulties on `xarray's issue tracker <https://github.com/pydata/xarray/issues>`_.
    In particular see discussion on `xarray issue #6807 <https://github.com/pydata/xarray/issues/6807>`_

Xarray can wrap chunked dask arrays (see :ref:`dask`), but can also wrap any other chunked array type that exposes the correct interface.

> This allows us to support using other frameworks for distributed and out-of-core processing, with user code still written as xarray commands.

Reviewer: Mention dask and cubed as explicitly supported up top.

The basic idea is that by wrapping an array that has an explicit notion of ``.chunks``, xarray can expose control over
the choice of chunking scheme to users via methods like :py:meth:`DataArray.chunk` whilst the wrapped array actually
implements the handling of processing all of the chunks.

Chunked array requirements
~~~~~~~~~~~~~~~~~~~~~~~~~~

> A chunked array needs to meet all the :ref:`requirements for normal duck arrays <internals.duckarrays>`, but should also implement these methods:

Reviewer: ```suggestion
A chunked array needs to meet all the :ref:`requirements for normal duck arrays <internals.duckarrays>`, but should also implement the following:
```

Reviewer: Can we link to the applicable section instead? Also there isn't a strict list [on that  page](https://docs.xarray.dev/en/latest/user-guide/duckarrays.html). So perhaps we should mention `from_array` here.

> - ``.chunk``

Reviewer: do you mean the `.chunks` property? What about the "core operations" below? Can we add an explicit list of which operations are required for which parts of Xarray's API?

Author: Yes I did mean that.

So what I've done now is get RST to actually render all the attributes and methods of the `ChunkManagerEntrypoint` ABC. That way I don't have to duplicate information, I can just improve the docstrings of those methods. This should also help to avoid future inconsistencies between the docs and the implementation.

Author: I can't seem to get sphinx to fully render all of the docstrings, not just the one-line summary of them. It is possible because it's done in [dask.distributed's API docs page](https://distributed.dask.org/en/latest/api.html#client) ([source](https://github.com/dask/distributed/blob/main/docs/source/api.rst#client)), but I can't see how the directives I've used are any different. Is there some kind of config command that might affect this?

- ``.rechunk``
- ``.compute``

Chunked operations as "core operations"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Xarray dispatches chunk-aware computations across arrays using "core operations" that accept one or more arrays.
Examples include ``map_blocks``, ``blockwise``, and ``apply_gufunc``.

> These core operations are generalizations of functions first implemented in :py:class:`dask.array`.

Reviewer: ```suggestion
These core operations are generalizations of functions first implemented in :py:module:`dask.array`.
```

Maybe?

The implementation of these functions is specific to the type of arrays passed to them: :py:class:`dask.array.Array` objects

> must be processed by :py:func:`dask.array.map_blocks`, whereas :py:class:`cubed.Array` objects must be processed by :py:func:`cubed.map_blocks`.

Reviewer: Need to add `cubed` here: https://github.com/pydata/xarray/blob/77e2db5aa1d8828bc7310062680ad9fe15a65d8a/doc/conf.py#L315

In order to use the correct implementation of a core operation for the array type encountered, xarray dispatches to the
corresponding subclass of :py:class:`~xarray.core.parallelcompat.ChunkManagerEntrypoint`,
also known as a "Chunk Manager". Therefore a full list of the primitive functions that need to be defined is set by the API of the
:py:class:`~xarray.core.parallelcompat.ChunkManagerEntrypoint` abstract base class.

.. note::

    The :py:class:`~xarray.core.parallelcompat.ChunkManagerEntrypoint` abstract base class is mostly just acting as a
    namespace for containing the chunked-aware function primitives. Ideally in the future we would have an API standard
    for chunked array types which codified this structure, making the entrypoint system unnecessary.

> ChunkManagerEntrypoint subclassing

Reviewer: Move this two paras up (including the note), that is also a bit redundant with the next line

Author: I'm not entirely sure what you mean here.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Rather than hard-coding various chunk managers to deal with specific chunked array implementations, xarray uses an
entrypoint system to allow developers of new chunked array implementations to register their corresponding subclass of
:py:class:`~xarray.core.parallelcompat.ChunkManagerEntrypoint`.

The key internal API is:

.. autosummary::

>    xarray.core.parallelcompat.list_chunkmanagers

Reviewer: I think you need to add these things to `api-hidden.rst`? The links aren't working...\

~~~~~~~~~~~~~~

Once the chunkmanager subclass has been registered, xarray objects wrapping the desired array type can be created in 3 ways:

#. By manually passing the array type to the :py:class:`DataArray` constructor, see the examples for :ref:`numpy-like arrays <userguide.duckarrays>`,

#. Calling :py:meth:`DataArray.chunk`, passing the keyword arguments ``chunked_array_type`` and ``from_array_kwargs``,

#. Calling :py:func:`open_dataset`, passing the keyword arguments ``chunked_array_type`` and ``from_array_kwargs``.

The latter two methods ultimately call the chunkmanager's implementation of ``.from_array``, to which they pass the ``from_array_kwargs`` dict.
The ``chunked_array_type`` kwarg selects which registered chunkmanager subclass to dispatch to. It defaults to ``'dask'`` if found,
otherwise to whichever chunkmanager is registered if only one is registered, else it will raise an error.

Parallel processing without chunks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To use a parallel array type that does not expose a concept of chunks explicitly, none of the information on this page
is theoretically required. Such an array type could be wrapped using xarray's existing

> support for :ref:`numpy-like "duck" arrays <userguide.duckarrays>`.

Reviewer: Nice add!

Shall we mention `arkouda`, `ramba` here
